x-pack/plugins/infra/public/pages/logs/settings/log_columns_configuration_panel.tsx

```diff
@@ -24,36 +23,61 @@ import { i18n } from '@kbn/i18n';
 import { FormattedMessage } from '@kbn/i18n/react';
 import React, { useCallback } from 'react';
 import { DragHandleProps, DropResult } from '../../../../../observability/public';
-import { LogColumnConfiguration } from '../../../utils/source_configuration';
-import { AddLogColumnButtonAndPopover } from './add_log_column_popover';
 import {
-  FieldLogColumnConfigurationProps,
-  LogColumnConfigurationProps,
-} from './log_columns_configuration_form_state';
+  FieldLogColumnConfiguration,
+  getLogColumnConfigurationId,
+  isMessageLogColumnConfiguration,
+  isTimestampLogColumnConfiguration,
+  LogColumnConfiguration,
+  MessageLogColumnConfiguration,
+  TimestampLogColumnConfiguration,
+} from '../../../utils/source_configuration';
+import { AddLogColumnButtonAndPopover } from './add_log_column_popover';
+import { FormElement } from './form_elements';
+import { FormValidationError } from './validation_errors';
 
-interface LogColumnsConfigurationPanelProps {
+export const LogColumnsConfigurationPanel = React.memo<{
   availableFields: string[];
   isLoading: boolean;
-  logColumnConfiguration: LogColumnConfigurationProps[];
-  addLogColumn: (logColumn: LogColumnConfiguration) => void;
-  moveLogColumn: (sourceIndex: number, destinationIndex: number) => void;
-}
+  logColumnsFormElement: FormElement<LogColumnConfiguration[], FormValidationError>;
+}>(({ availableFields, isLoading, logColumnsFormElement }) => {
+  const addLogColumn = useCallback(
```

Comment:
:information_source: Previously these concerns were mixed into the form state management. Sorry for the noise here, but this way the knowledge is less spread out.